toxcore/network.c

```diff
@@ -134,7 +138,12 @@ static int inet_pton6(const char *addrString, struct in6_addr *addrbuf)
 #define IPV6_V6ONLY 27
 #endif
 
-#define TOX_EWOULDBLOCK WSAEWOULDBLOCK
+static bool should_ignore_recv_error(int err)
+{
+    // We ignore WSAECONNRESET as Windows helpfully* sends that error if a
+    // previously sent UDP packet wasn't delivered.
```

Comment:
Wait a minute, how does Windows know if a packet was delivered or not? UDP is stateless, there is no delivery ack that the other party sends to us upon delivery to let us know they received the packet.

Reply:
The way I understood it is that the remote responds with an ICMP unreachable message if nothing's listening on the destination host/port. :x The documentation isn't great on this, but it seems to be a common issue with UDP sockets on Windows.

Reply:
After googling around for a bit it all makes sense:

>[On a UDP-datagram socket this error indicates a previous send operation resulted in an ICMP Port Unreachable message.](https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-recvfrom?redirectedfrom=MSDN)


These comments were also very useful:
>This is an "optimization" done by Microsoft.
>
>When you sendto() to a host/port combination that isn't actually expecting a UDP datagram, that host will likely send back a ICMP error message.
>
>In the regular socket API, there isn't a good way of actually receiving this error, so Microsoft wired that message back up to the call to recvfrom(). If there is a received, un-delivered ICMP error of this kind, it will be returned by recvfrom(). If you want to know which particular remote host/port didn't respond, check the address returned by recvfrom().
>
>This is actually useful. You can use this information to detect that a remote machine is not running a server (if this is on the client,) or that a remote client has shut down / crashed (if this is on the server.) Or you can just use it to detect that you're passing the wrong parameters to sendto() and are not sending the datagrams to the address/port you think you are :-)
>
>The main thing to realize about this error is that it doesn't mean the socket is dead -- you can loop right back and call recvfrom() on it again, to dequeue whatever UDP datagrams are pending on the socket.

https://gamedev.net/forums/topic/645149-recvfrom-and-wsaeconnreset/5075595/

>Yes, it's faster than a timeout, but it's not guaranteed. Some hosts don't send the ICMP; some networks filter them; some machines may quit by being turned off (not being able to respond at all) etc. So you need the timeout, and can perhaps accelerate detection of some disconnects with this message.

https://gamedev.net/forums/topic/645149-recvfrom-and-wsaeconnreset/5075625/